_posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd

```diff
@@ -0,0 +1,347 @@
+*has* to peek into things for them to make sense.
+
+#### Digging deeper
+
+We build up a simple graph of tensors, with "inputs" `x1` and `x2` being
+connected to "output" `out` by "intermediaries" `y` and `z`.
+
+```{r}
+x1 <- torch_ones(2, 2, requires_grad = TRUE)
+x2 <- torch_tensor(1.1, requires_grad = TRUE)
+
+y <- x1 * (x2 + 2)
+
+z <- y$pow(2) * 3
+
+out <- z$mean()
+```
+
+To save memory, by default intermediate gradients are not stored.
+Calling `retain_grad()` on a tensor allows to deviate from this default;
+let's do this here, for the sake of demonstration:
+
+```{r}
+y$retain_grad()
+
+z$retain_grad()
+```
+
+Now we can go backwards through the graph and inspect `torch`'s action
+plan for backprop, starting from `out$grad_fn`, like so:
+
+```{r}
+# how to compute the gradient for mean, the last operation executed
+out$grad_fn
+```
+
+    MeanBackward0
+
+```{r}
+# how to compute the gradient for the multiplication by 3 in z = y.pow(2) * 3
+out$grad_fn$next_functions
+```
+
+    [[1]]
+    MulBackward1
+
+```{r}
+# how to compute the gradient for pow in z = y.pow(2) * 3
+out$grad_fn$next_functions[[1]]$next_functions
+
+```
+
+    [[1]]
+    PowBackward0
+
+```{r}
+# how to compute the gradient for the multiplication in y = x * (x + 2)
+out$grad_fn$next_functions[[1]]$next_functions[[1]]$next_functions
+
+```
+
+    [[1]]
+    MulBackward0
+
+```{r}
+# how to compute the gradient for the two branches of y = x * (x + 2),
+# where the left branch is a leaf node (AccumulateGrad for x1)
+out$grad_fn$next_functions[[1]]$next_functions[[1]]$next_functions[[1]]$next_functions
+
+```
+
+    [[1]]
+    torch::autograd::AccumulateGrad
+    [[2]]
+    AddBackward1
+
+```{r}
+# here we arrive at the other leaf node (AccumulateGrad for x2)
+out$grad_fn$next_functions[[1]]$next_functions[[1]]$next_functions[[1]]$next_functions[[2]]$next_functions
+
+```
+
+    [[1]]
+    torch::autograd::AccumulateGrad
+
+If we now call `out$backward()`, all tensors in the graph will have
+their respective gradients calculated.
+
+```{r}
+out$backward()
+
+z$grad
+y$grad
+x2$grad
+x1$grad
+```
+
+    torch_tensor 
+     0.2500  0.2500
+     0.2500  0.2500
+    [ CPUFloatType{2,2} ]
+    torch_tensor 
+     4.6500  4.6500
+     4.6500  4.6500
+    [ CPUFloatType{2,2} ]
+    torch_tensor 
+     18.6000
+    [ CPUFloatType{1} ]
+    torch_tensor 
+     14.4150  14.4150
+     14.4150  14.4150
+    [ CPUFloatType{2,2} ]
+
+After this nerdy excursion, let's see how *autograd* makes our network
+simpler.
+
+## The simple network, now using *autograd*
+
+Thanks to *autograd*, we say good-bye to the tedious, error-prone
+process of coding backpropagation ourselves. A single method call does
+it all: `loss$backward()`.
+
+With `torch` keeping track of operations as required, we don't even have
+to explicitly name the intermediate tensors any more. Just for fun and
+to make the point, we can code forward pass, loss calculation and
+backward pass in just three lines:
```

Comment:
I think this point can be made without the qualifier (especially since you have another "just" in the sentence)…also allows you to use an Oxford comma in the list, which is nice because "pass" is a noun and a verb.

Reply:
that is excellent!! I'm being overly explicit sometimes it seems